Adds `unmanaged-cluster` TKR compatibility check

jpmcb opened the pull request:
## What this PR does / why we need it
- Queries registry for list of tags for compatibility file. Uses the most recently major versioned tag. I.e., if there are `[v1, v2, v3]` tags, it will select `v3` since it's the "latest" tag
- Checks if the tkr compatibility file is already downloaded. If it doesn't exist, downloads and unpacks it from the registry.
- Reads and unmarshals compatibility file
  - Assumes unmarshaled yaml is ordered. Selects first compatible TKr. PR includes test for this
- Checks version of unmanaged-cluster CLI plugin vs compatible TKR. Uses (and downloads if needed) tkr that corresponds to a compatible version.
- If user provides a tkr via `--tkr` flag, checks to see if it's in the compatibility file. If not, logs a warning.
- Helper script for pushing a compatibility file. Sample compatibility file also in PR
- Moves getting user package repo out of the `config` package and into the TKr.
  - If a user provides their own `additional-repo` via flags or configuration, uses that instead

## Which issue(s) this PR fixes
Fixes: #3538

## Describe testing done for PR
Compatibility file in registry:
```
$ crane ls projects.registry.vmware.com/tce/compatibility
v1
v2
v3
```

File contents:
```yaml
version: v1
unmanagedClusterPluginVersions:
- version: dev
  supportedTkrVersions:
  - image: projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
  - image: projects.registry.vmware.com/tce/tkr:v0.17.0
  - image: projects.registry.vmware.com/tce/tkr:v1.22.5
- version: v0.11.0
  supportedTkrVersions:
  - image: projects.registry.vmware.com/tce/tkr:v0.17.0
  - image: projects.registry.vmware.com/tce/tkr:v1.22.5
- version: v0.10.0
  supportedTkrVersions:
  - image: projects.registry.vmware.com/tce/tkr:v0.21.5
```

The TKr now should have the following section:
```yaml
  tkg-core-packages:
  - version: v1.21.2+vmware.1-tkg.1
    images:
      tanzuCorePackageRepositoryImage:
        imagePath: repo-10
        tag: 0.10.0
        repository: projects.registry.vmware.com/tce
      tanzuUserPackageRepositoryImage:
        imagePath: main
        tag: 0.11.0
        repository: projects.registry.vmware.com/tce
```

and is present in `projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1` which is at the top of the compatibility file for the `dev` version

---

Creating a cluster, selects the correct TKR
```
$ tanzu unmanaged-cluster create test

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Downloaded to: /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
   Downloaded to: /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0-dev-1
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/test/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/test/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:v0.11.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1

🚀 Creating cluster test
   Cluster creation using kind!
   ❤️  Checkout this awesome project at https://kind.sigs.k8s.io
   Base image downloaded
   Cluster created
   To troubleshoot, use:

   kubectl ${COMMAND} --kubeconfig /home/jmcb/.config/tanzu/tkg/unmanaged/test/kube.conf

📧 Installing kapp-controller
   kapp-controller status: Running

📧 Installing package repositories
   Core package repo status: Reconcile succeeded

🌐 Installing CNI
   calico.community.tanzu.vmware.com:3.22.1

✅ Cluster created

🎮 kubectl context set to test

View available packages:
   tanzu package available list
View running pods:
   kubectl get po -A
Delete this cluster:
   tanzu unmanaged delete test
```

Running create again skips downloading compatibility file since it already exists

```
$ tanzu unmanaged-cluster create test-2

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
   TKr exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0-dev-1
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/test/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/test/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected additional package repositories
   projects.registry.vmware.com/tce/main:0.11.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1
...
```

Using the `--tkr` flag works and checks to ensure it's in the compatibility file 👍🏼 
```
$ tanzu unmanaged-cluster create test --tkr projects.registry.vmware.com/tce/tkr:v1.22.5

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3
   Custom Tkr projects.registry.vmware.com/tce/tkr:v1.22.5 found in compatibility file

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v1.22.5
   TKr exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.22.5
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/test-3/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/test-3/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind/node:v1.22.5

📦 Selected core package repository
   projects-stg.registry.vmware.com/tkg/packages/core/repo:v1.22.5_vmware.1-tkg.3-tf-v0.11.2

📦 Selected kapp-controller image bundle
   projects-stg.registry.vmware.com/tkg/packages/core/kapp-controller:v0.30.0_vmware.1-tkg.1
...
```

If the tkr is _not_ in the compatibility file, it raises a warning, but proceeds (fyi, this fails to download tkr since it doesn't exist, but demonstrates tkr warning if not in compatibility file):
```
$ tanzu unmanaged-cluster create test --tkr projects.registry.vmware.com/tce/tkr:v1.99.99

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3
   Custom TKr projects.registry.vmware.com/tce/tkr:v1.99.99 NOT found in compatibility file. Proceed with caution, the provided TKr may not work with this version of unmanaged-cluster

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v1.99.99
   Downloading to: /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v1.99.99
...
```


Looking at the file system in `~/.config/tanzu/tkg/unmanaged/`:
```
❯ tree
.
├── bom
│   ├── projects.registry.vmware.com_tce_tkr_v0.17.0-dev-1
│   ├── projects.registry.vmware.com_tce_tkr_v0.17.0
│   └── projects.registry.vmware.com_tce_tkr_v1.22.5
├── compatibility
│   └── projects.registry.vmware.com_tce_compatibility_v3
├── test
│   ├── bootstrap.log
│   ├── config.yaml
│   └── kindconfig.yaml
```

Providing your _own_ user managed repo works as expected and will _not_ select the one in the downloaded tkr:
```
$ tanzu unmanaged-cluster create woof-test --additional-repo projects.registry.vmware.com/main:v0.10.0

📁 Created cluster directory

🧲 Resolving and checking Tanzu Kubernetes release (TKr) compatibility file
   projects.registry.vmware.com/tce/compatibility
   Compatibility file exists at /home/jmcb/.config/tanzu/tkg/unmanaged/compatibility/projects.registry.vmware.com_tce_compatibility_v3

🔧 Resolving TKr
   projects.registry.vmware.com/tce/tkr:v0.17.0-dev-1
   TKr exists at /home/jmcb/.config/tanzu/tkg/unmanaged/bom/projects.registry.vmware.com_tce_tkr_v0.17.0-dev-1
   Rendered Config: /home/jmcb/.config/tanzu/tkg/unmanaged/woof-test/config.yaml
   Bootstrap Logs: /home/jmcb/.config/tanzu/tkg/unmanaged/woof-test/bootstrap.log

🔧 Processing Tanzu Kubernetes Release

🎨 Selected base image
   projects.registry.vmware.com/tce/kind:v1.22.4

📦 Selected core package repository
   projects.registry.vmware.com/tce/repo-10:0.10.0

📦 Selected additional package repositories
   projects.registry.vmware.com/main:v0.10.0

📦 Selected kapp-controller image bundle
   projects.registry.vmware.com/tce/kapp-controller-multi-pkg:v0.30.1
```

Note that if a user provides a tkr that does _not_ have a user managed repo in it, we won't attempt to select resolve any kind of default and it won't be logged as being selected.

jpmcb commented:
A few thoughts we might want to consider:
- Logging a message if the user uses the `--tkr` flag to warn that checking the compatability file will be skipped
- Refactor and make the use of `projects.registry.vmware.com/tce` more dynamic. We are using that string all over the place and we may want to (in the future or this PR) enable users to configure their registry url

Things I'm unsure of
- How do we handle multiple compatible TKRs for a version of the `unmanaged-cluster` plugin. Right now, I've just punted to take the first one. But this should be smarter and less brittle.
  - On that note, is it possible we will ever ship a version with _no_ compatible tkr? My guess is no, but to prevent a panic, we should probably put a check in to prevent out of bounds when selecting the tkr
- Is "Tanzu Kubernetes Release" supposed to be `TKr`? Or `TKR`?

joshrosso reviewed (changes requested):
just marking myself to review.

thanks :rocket:

joshrosso reviewed (commented):
> Refactor and make the use of projects.registry.vmware.com/tce more dynamic. We are using that string all over the place and we may want to (in the future or this PR) enable users to configure their registry url

Regarding the compatibility file:

```
version: v1
unmanagedClusterPluginVersions:
- version: dev
  supportedTkrVersions:
  - imagePath: tkr
    tag: v0.17.0
- version: v0.11.0
  supportedTkrVersions:
  - imagePath: tkr
    tag: v0.17.0
- version: v0.10.0
  supportedTkrVersions:
  - imagePath: tkr
    tag: v1.21.5
```

I'd like to simplify the schema to:

```
version: v1
unmanagedClusterPluginVersions:
- version: dev
  supportedTkrVersions:
  - image: projects.registry.vmware.com/tce/tkr:v0.17.0
  - image: projects.registry.vmware.com/tce/tkr:v1.22.5
- version: v0.11.0
  supportedTkrVersions:
  - image: projects.registry.vmware.com/tce/tkr:v0.17.0
  - image: projects.registry.vmware.com/tce/tkr:v1.22.5
```



Reasons being:

* Trying to be clever with our image path resolution cause increased code complexity.
* In TKG, it's been really tough to know where to modify things like image repo, because it's not explicit.
* Using the syntax `image:` enables use to use a common paradigm, so that tools like `imgpkg` could copy this file and all associated assets to a local registry.

Sorry I didn't include this in the original issue/spec. But I think we can learn something from the design of TKG here and improve.

joshrosso commented:
> Logging a message if the user uses the --tkr flag to warn that checking the compatibility file will be skipped

I like this. In conjunction with https://github.com/vmware-tanzu/community-edition/pull/3641#pullrequestreview-919553343, I think we should check whether the `--tkr` specified is in the list of known-compatible TKrs for the plugin version.

If it is not, **then** we should print the warning you describe.

joshrosso commented:
> How do we handle multiple compatible TKRs for a version of the unmanaged-cluster plugin. Right now, I've just punted to take the first one. But this should be smarter and less brittle.

To the best of my knowledge, YAML arrays, unlike maps are ordered:

https://yaml.org/spec/1.2.2/#311-dump

The implication of this is our unmarshaling library should preserve order when converting to a go slice. We should verify this, but if it is guaranteed, we should prefer index `[0]` as the default.

> On that note, is it possible we will ever ship a version with no compatible tkr? My guess is no, but to prevent a panic, we should probably put a check in to prevent out of bounds when selecting the tkr

No. But yes, we should not panic due to the exceptional case where a compatibility file is malformed. In this case, we should raise an `error`.

joshrosso commented:
> Is "Tanzu Kubernetes Release" supposed to be TKr? Or TKR?

Tanzu Kubernetes release is supposed to be TKr.

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -26,21 +26,25 @@ import (
+	tkgGlobalPkgNamespace    = "tanzu-package-repo-global"
+	tceCompatibilityRegistry = "projects.registry.vmware.com/tce/compatibility"
+	tceRepoName              = "community-repository"
+	tceRepoURL               = "projects.registry.vmware.com/tce/main:v0.11.0"
```
As part of this work, I'd like to see the user-managed package repo be put in the TKr going forward.

For example, here's an updated version of our TKr.

```yaml
  tkg-core-packages:
  - version: v1.21.2+vmware.1-tkg.1
    images:
      tanzuCorePackageRepositoryImage:
        imagePath: repo-10
        tag: 0.10.0
        repository: projects.registry.vmware.com/tce
      tanzuUserPackageRepositoryImage:
        imagePath: main
        tag: 0.11.0
        repository: projects.registry.vmware.com/tce
```

joshrosso replied:
To test this out, feel free to work off a new TKr you create like `project.registry.vmware.com/tce/tkr/v0.17.0-dev+1`

jpmcb replied:
Hmmmm I think that the `tceRepoUrl` const in the `tanzu` package was some leftover from work done to make `additionalPackageRepos` a configurable option via the config.

So how should that play in here? A few thoughts:
- The config by default is empty and it pulls from the TKr instead
- What would the behavior be if the user provides their own `additionalPackageRepos` in a config file or via flag?
  - My guess would be it ignores the user managed package repo in the TKr and uses the ones provided by the user

https://github.com/vmware-tanzu/community-edition/blob/7132015da9057fa45d0c36178d7651478de0599d/cli/cmd/plugin/unmanaged-cluster/config/config.go#L54-L56

joshrosso replied:
good points. let's talk this one through synchronously today.

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -451,6 +472,118 @@ func createClusterDirectory(clusterName string) (string, error) {
+		}
+	}
+
+	return "", fmt.Errorf("could not find compatible CLI version in compatibility file")
```
The way this bubbles up, it gives the user a bit of a confusing error message. Namely, it says:

![image](https://user-images.githubusercontent.com/6200057/159814083-1367f09d-2845-48b7-a19e-2190b9ea26c8.png)

Is there a way (without too much complex logic) we could tell the user:

"Failed parsing the TKr compatibility file: Error...."

joshrosso commented on cli/cmd/plugin/unmanaged-cluster/tanzu/tanzu.go:
```diff
@@ -135,7 +136,18 @@ func (t *UnmanagedCluster) Deploy(scConfig *config.UnmanagedClusterConfig) (int,
+	// Download compatibility file and read TKr version compatible with version of unmanaged-cluster CLI
+	// when the user did _not_ set the TKr via a flag or configuration option
+	if scConfig.TkrLocation == "" {
+		log.Event(logger.MagnetEmoji, "Resolving and checking TKr compatibility file")
```
This will be the **new** first time we mention TKr to the user. Can we spell it out here, and use the abbreviation in subsequent logs?

![image](https://user-images.githubusercontent.com/6200057/159814352-a03bf43c-be1e-4e8b-b4da-d860381f52ce.png)

joshrosso reviewed (changes requested):
This is looking awesome :rocket: .

Suggestions made throughout. Will look for responses/updates tomorrow.

jpmcb commented:
Pushed a second commit:
- Updated to reflect new, simpler scheme. Should now pull `projects.registry.vmware.com/tce/compatibility:v2`
- Added a test to ensure ordering of slices are consistent and ordered
- When users provides `--tkr` flag, will check against compatibility file to see if it's in list of compatible TKrs. If not, will log a warning.
- Added simple script for pushing a compatibility file and a sample compatibility file in the `hack` directory. If we don't want these, can remove
- Updated top level PR description to reflect changes

jpmcb committed:
Adds unmanaged-cluster TKr compatibility check

- Checks for a local tkr compatibility file. If it doesn't exist,
  download it from the registry
- Checks version of CLI vs compatible TKr. Uses (and downloads if
  needed) the latest tkr that correspodns to a compatible version
- Adds test to ensure ordering of campatibility versions: the "latest"
  should always be first
- Removes the user package repository registry location from the config
  and moves it to the TKr

Signed-off-by: John McBride <jmcbride@vmware.com>

jpmcb commented:
This is ready to go. Updated the top level description with new behaviors and new field in TKr for user package repo.

The lint check appears to be unrelated. Ran `golangci-lint run -c ../../../../.golangci.yaml` in the unmanaged directory and it succeeds.

joshrosso reviewed (changes requested):
@jpmcb one small change and then we're ready.

When the UserManaged Package repo isn't there:

![image](https://user-images.githubusercontent.com/6200057/160498898-c227ad01-d7f8-4ea5-9019-f38a77b0413a.png)

Can we not log this at all?

jpmcb commented:
> Can we not log this at all?

Done! ✅
